Test Conway CDDL using Cuddle

nc6 opened the pull request:
# Description

This PR introduces [cuddle](https://github.com/input-output-hk/cuddle), a Haskell library for declaring and validating CDDL documents.

This is then used to reflect the Conway CDDL specification (in eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs) using a DSL known as Huddle.

libs/cardano-ledger-binary/testlib/Test/Cardano/Ledger/Binary/Cuddle.hs introduces utilities to run CDDL conformance tests using cuddle. These are currently based on a Huddle declaration, but it would be equally possible to use a native CDDL specification combined with cuddle's CBOR generators.

eras/conway/impl/test/Test/Cardano/Ledger/Conway/Binary/CddlSpec.hs then introduces a new set of tests based in the Huddle conway specification. The existing test suite is left untouched for now.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

nc6 committed:
Test Conway round-trip with Cuddle

This commit introduces a re-implementation of the
Conway CDDL specification using 'cuddle', a
Haskell library for defining and validating CDDL.
This allows us to generate and validate test cases
without resorting to using the external `cddl`
library.

nc6 committed:
Ensure reproducibility of test cases.

nc6 committed:
Various fixes to the Cuddle Conway specification:

- Fix protocol_param_update definition
- Fix indices for certificates.
- Inputs to a transaction should be inputs, not indices.
- Index is sized
- Fix the alternatives in credential.
- Fix tags for constr.
- script_ref has a tag
- post_alonzo_transaction_output is a map
- voter and gov_action_id are refs, not keywords
- Appropriate size the index tag in redeemers
- Both sides of the redeemers map are arrays, not maps

nc6 committed:
Actually run multiple examples.

nc6 committed:
Fix CDDL representing CostModels in Conway.

Reflects https://github.com/IntersectMBO/cardano-ledger/pull/4390

nc6 committed:
Run cabal-format.

nc6 committed:
Update to latest cuddle version

In particular, updating to this version prevents generating CBOR Maps
with duplicate keys, which can result in parsing errors

teodanciu commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/Binary/CddlSpec.hs:
```diff
@@ -15,28 +16,59 @@ import Test.Cardano.Ledger.Binary.Cddl (
+  huddleRoundTripAnnCborSpec @(TxWits Conway) v "transaction_witness_set"
+  huddleRoundTripAnnCborSpec @(Redeemers Conway) v "redeemers"
+  huddleRoundTripAnnCborSpec @(Tx Conway) v "transaction"
+  pure ()
```
Seems unnecessary: 
```suggestion
```

nc6 replied:
Whoops. Yeah, that was a hangover from testing when I was commenting things out. Will fix

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,971 @@
+-- The real type of  plutus_v1_script, plutus_v2_script and plutus_v3_script is bytes.
+-- However, because we enforce uniqueness when many scripts are supplied,
+-- we need to hack around for tests in order to avoid generating duplicates,
+-- since the cddl tool we use for roundtrip testing doesn't generate distinct collections.
```
Is this comment still applicable? it seems like `distinct` has solved the problem that's being described in, in which case we can get rid of it.
```suggestion
```

nc6 replied:
Isn't `distinct` a hack, though? Precisely the hack that we're referring to.

teodanciu replied:
Sorry, I assumed we don't need the hack anymore, but indeed, since we still use distinct in the type itself, the comment makes sense.

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,971 @@
+        <> "we need to hack around for tests in order to avoid generating duplicates,\n"
+        <> "since the cddl tool we use for roundtrip testing doesn't generate distinct collections.\n"
+    )
+    $ "plutus_v1_script" =:= distinct VBytes
```
```suggestion
   "plutus_v 1_script" =:= distinct VBytes
```

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,971 @@
+import GHC.Num (Integer)
+import GHC.Show (Show (show))
+
+default (Integer, Double)
```
Isn't this the default default?  I think we could get rid of this :thinking: 
```suggestion
```

nc6 replied:
Yes, it never helped anyway

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,971 @@
+conway :: Huddle
+conway = collectFrom [block, transaction]
+
+block :: Rule
```
Is the plan for the future to define these in Shelley and to reuse them here (where appropriate)?   And like this to replace our cddl-tool that generates examples?

nc6 replied:
Yes, this should all eventually be modularised - that's the next thing on my TODO list!

nc6 replied:
And we can already generate examples without using the cddl-tool!

teodanciu replied:
Aha, I meant we can remove the tool once we have the definitions for all eras. Nice!

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,971 @@
+      , "issuer_vkey" ==> vkey
+      , "vrf_vkey" ==> vkey
+      , "vrf_result" ==> vrf_cert
+      , "block_body_size" ==> VUInt
```
Are we using `VUInt` to represent all sizes?  Is this the same VUint defined like this in cuddle?
```
data Value
  = VUInt Word64
  | VNInt Word64
   ..
```

How does it work then,  can we define precise sizes for the unsigned integers as we do in the cddl file?

lehins replied:
This is a very important question. We do need ability to specify sizes, eg. [`block_body_size  : uint .size 4`](https://github.com/IntersectMBO/cardano-ledger/blob/f4119c9cd27e69943dcf91d22d0f9905be32fa75/eras/conway/impl/cddl-files/conway.cddl#L34)

I assume it can be achieved with something like this, right?
```suggestion
      , "block_body_size" ==> VUInt `sized` (4 :: Word)
```

nc6 replied:
Yes, the `sized` function works exactly in this way. Updated to reflect this

teodanciu commented on libs/cardano-ledger-binary/cardano-ledger-binary.cabal:
Do you mind updating the package version  to 1.3.4.0 ?

teodanciu commented on eras/conway/impl/cardano-ledger-conway.cabal:
```diff
@@ -96,6 +97,7 @@ library
         cardano-slotting,
         cardano-strict-containers,
         containers,
+        cuddle,
```
`cardano-ledger-binary` version above needs to be updated to >= 1.3.4  (if  [this suggestion is followed](https://github.com/IntersectMBO/cardano-ledger/pull/4437/files#r1662738083))

lehins replied:
We don't want libraries to depend on `cuddle`, because neither that library not the CDDL spec that is generated from it is not necessary for cardano-node execution. We only need `cuddle` for testing and producing CDDL spec, so everything `cuddle` related that is reusable should go into `testlib`.

nc6 replied:
It already is in testlib. Should one update the version when the testlib library is updated but nothing else?

nc6 replied:
Oh, sorry, you mean that Conway depends on Cuddle. I can move to testlib, I had placed it in the library merely to reflect the current location of the CDDL

lehins replied:
I'll be a bit more precise. This is where it needs to go: https://github.com/IntersectMBO/cardano-ledger/blob/8cff9cfcf23686bbde8bfa5dc0682630af38c517/eras/conway/impl/cardano-ledger-conway.cabal#L112

lehins replied:
Current location of CDDL is in [the package](https://github.com/IntersectMBO/cardano-ledger/blob/8cff9cfcf23686bbde8bfa5dc0682630af38c517/eras/conway/impl/cardano-ledger-conway.cabal#L16), not the library. So, sublibraries have full access to those files as well

lehins commented on eras/conway/impl/test/Test/Cardano/Ledger/Conway/Binary/CddlSpec.hs:
```diff
@@ -15,28 +16,59 @@ import Test.Cardano.Ledger.Binary.Cddl (
+        cddlRoundTripCborSpec @(VotingProcedure Conway) v "voting_procedure"
+        cddlRoundTripCborSpec @(ProposalProcedure Conway) v "proposal_procedure"
+        cddlRoundTripCborSpec @(GovAction Conway) v "gov_action"
+        cddlRoundTripCborSpec @(TxCert Conway) v "certificate"
```
```haskell
foo $ 
  do
    bar
```

This is the style we try to avoid in ledger :wink: Also, this diff is totally unnecessary, so let's minimize it by reverting the change.
```suggestion
  describe "CDDL" $ beforeAllCddlFile 3 readConwayCddlFiles $ do
    let v = eraProtVerHigh @Conway
    cddlRoundTripCborSpec @(Value Conway) v "positive_coin"
    cddlRoundTripCborSpec @(Value Conway) v "value"
    cddlRoundTripAnnCborSpec @(TxBody Conway) v "transaction_body"
    cddlRoundTripAnnCborSpec @(TxAuxData Conway) v "auxiliary_data"
    cddlRoundTripAnnCborSpec @(Timelock Conway) v "native_script"
    cddlRoundTripAnnCborSpec @(Data Conway) v "plutus_data"
    cddlRoundTripCborSpec @(TxOut Conway) v "transaction_output"
    cddlRoundTripAnnCborSpec @(Script Conway) v "script"
    cddlRoundTripCborSpec @(Datum Conway) v "datum_option"
    cddlRoundTripAnnCborSpec @(TxWits Conway) v "transaction_witness_set"
    cddlRoundTripCborSpec @(PParamsUpdate Conway) v "protocol_param_update"
    cddlRoundTripCborSpec @CostModels v "costmdls"
    cddlRoundTripAnnCborSpec @(Redeemers Conway) v "redeemers"
    cddlRoundTripAnnCborSpec @(Tx Conway) v "transaction"
    cddlRoundTripCborSpec @(VotingProcedure Conway) v "voting_procedure"
    cddlRoundTripCborSpec @(ProposalProcedure Conway) v "proposal_procedure"
    cddlRoundTripCborSpec @(GovAction Conway) v "gov_action"
    cddlRoundTripCborSpec @(TxCert Conway) v "certificate"
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,971 @@
+update_drep_cert = "update_drep_cert" =:~ grp [18, a drep_credential, anchor // VNil]
+
+delta_coin :: Rule
+delta_coin = "delta_coin" =:= VUInt
```
`delta_coin` is actually signed
```haskell
delta_coin = "delta_coin" =:= VInt
```

But, it is unused so we can get rid of it. It is unnecessary in the original spec either, so if you can remove it there too that would be helpful.
```suggestion
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,971 @@
+    =:= arr
+      [ "block_number" ==> VUInt
+      , "slot" ==> VUInt
+      , "prev_hash" ==> hash32
```
`prev_hash` is optional
```suggestion
      , "prev_hash" ==> hash32 // VNil
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,971 @@
+      , "slot" ==> VUInt
+      , "prev_hash" ==> hash32
+      , "issuer_vkey" ==> vkey
+      , "vrf_vkey" ==> vkey
```
```suggestion
      , "vrf_vkey" ==> vrf_vkey
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,971 @@
+      ]
+
+protocol_version :: Named Group
+protocol_version = "protocol_version" =:~ grp [a major_protocol_version, a VUInt]
```
Can we define it like that, instead of using `(arr [a protocol_version])` at use sites?
```suggestion
protocol_version = "protocol_version" =:~ arr [a (grp [a major_protocol_version, a VUInt])]
```

This way we would be able to use it as before [`a protocol_version`](https://github.com/IntersectMBO/cardano-ledger/blob/f4119c9cd27e69943dcf91d22d0f9905be32fa75/eras/conway/impl/cddl-files/conway.cddl#L37)

nc6 replied:
Yes, this is fine. My translation of the Conway CDDL preceded 280ace3ed7, hence this change.

nc6 replied:
Though there's not really any point in the extra group indirection here.

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,971 @@
+next_major_protocol_version = "next_major_protocol_version" =:= (10 :: Integer)
+
+major_protocol_version :: Rule
+major_protocol_version = "major_protocol_version" =:= (1 :: Integer)
```
This should be a list from [`1..next_major_protocol_version`](https://github.com/IntersectMBO/cardano-ledger/blob/f4119c9cd27e69943dcf91d22d0f9905be32fa75/eras/conway/impl/cddl-files/conway.cddl#L49)
```suggestion
major_protocol_version = "major_protocol_version" =:= (1 :: Integer)
```

nc6 replied:
Ah, currently huddle doesn't support this. No reason it can't, but for now I've inlined it and added a TODO.

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,971 @@
+      [ idx 0 ==> set transaction_input
+      , idx 1 ==> arr [0 <+ a transaction_output]
+      , idx 2 ==> coin
+      , opt (idx 3 ==> VUInt)
```
We now have defined [`slot_no`](https://github.com/IntersectMBO/cardano-ledger/blob/f4119c9cd27e69943dcf91d22d0f9905be32fa75/eras/conway/impl/cddl-files/conway.cddl#L589)
```suggestion
      , opt (idx 3 ==> slot_no)
```

lehins reviewed (changes requested):
This looks great!!! I am loving it!!! :heart: 

Looks like you've used an older CDDL spec as reference, could you please go through and compare to an up-to-date version and make necessary adjustments?

I haven't gone through the whole PR, but as soon as all current comments are addressed I'll do a thorough review.

This PR pretty much takes care of #4125 and #4126

There are more to go: #4461 and #4462 and then some other ideas that we've discussed

nc6 committed:
Apply fixes from code review.

nc6 committed:
Bring Cuddle up to date with CDDL changes.

Updated with all changes to conway.cddl up to 78b32d5

nc6 committed:
Final changes via comparison with conway.cddl

This is the final set of changes to the Conway
CDDL spec made via comparison with the original
cddl files for Conway.

As of this commit, the remaining changes with the
original cddl are as follows:

- Various fields in crypto were defined with /=
  and are now defined with =. This does not affect
  the semantics, being instead a convention as per
  https://datatracker.ietf.org/doc/html/rfc8610#section-3.9
- The index keys in auxiliary_data are written as
  if they were value and not type keys:
  ```
    39,43c30,34
      <                   / #6.259({? 0 : metadata
      <                             , ? 1 : [* native_script]
      <                             , ? 2 : [* plutus_v1_script]
      <                             , ? 3 : [* plutus_v2_script]
      <                             , ? 4 : [* plutus_v3_script]})
      ---
      >                   / #6.259({? 0 => metadata
      >                             , ? 1 => [* native_script]
      >                             , ? 2 => [* plutus_v1_script]
      >                             , ? 3 => [* plutus_v2_script]
      >                             , ? 4 => [* plutus_v3_script]})
  ```
  This is a simpler formulation with identical semantics.
- The alternatives in 'certificate', 'relay',
  'native_script' and 'gov_action' are encoded
  using type1 choice semantics ('/') rather than
  group choice semantics ('//'). In the context,
  this is an equivalent formulation.
- nonempty_oset has been removed, pending
  explanation of what it means. nonempty_set is
  used instead.
- In various places, null is replaced with nil.
  They are synonyms.
- The names of type parameters are changed;
  typically a is replaced with a0.
- A few rules using references as the bounds in
  ranges are now inlined. This will be fixed once
  https://github.com/input-output-hk/cuddle/issues/29
  is addressed.
- `distinct` is not currently possible to
  implement as a generic function, since `sized`
  is currently specialised to work on values.
  https://github.com/input-output-hk/cuddle/issues/31
  addresses. Until this is fixed, distinct<bytes>
  is instead distinct_bytes.
- tstr is replaced with text. They are synonyms.
- The keys in the 'pool_params' and
  'script_n_of_k' groups are lost. This is due to
  https://github.com/input-output-hk/cuddle/issues/32.
  However, since pool_params is included into an
  array, this does not alter the semantics, only
  the documentation.
- datum_hash is of type datum_hash, not hash32.
  This seems an overlooked item in the original
  cddl.

nc6 committed:
Move Cuddle spec to the testlib.

nc6 commented:
Note that I have #4461 addressed, but planned to include that in a subsequent PR to avoid inflating this one.

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,1008 @@
+      , "transaction_bodies" ==> arr [0 <+ a transaction_body]
+      , "transaction_witness_sets"
+          ==> arr [0 <+ a transaction_witness_set]
+      , "auxiliary_data_set"
+          ==> mp [0 <+ asKey transaction_index ==> auxiliary_data]
```
This is interesting. In Huddle `==>` operator is used for two different types of syntax:
```
[ key : value ]
```
and
```
{ key => value }
```
Do they have a similar meaning in CDDL or was it just possible to do this for convenience in Huddle?

nc6 replied:
In CDDL the `key : value` syntax is short for `"key" ==> value`. Strings have to be quoted in Haskell, so they're merged.

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,1008 @@
+transaction_index = "transaction_index" =:= VUInt `sized` (2 :: Word64)
+
+header :: Rule
+header = "header" =:= arr [a header_body, "body_signature" ==> kes_signature]
```
It is a bit easier to see that those are separate fields this way
```suggestion
header = "header" =:= arr 
  [ a header_body
  , "body_signature" ==> kes_signature
  ]
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,1008 @@
+  "parameter_change_action"
+    =:~ grp
+      [ 0
+      , gov_action_id // VNil
```
CDDL is too confusing! So, in the current spec it is defined as `gov_action_id / null`, where it is one `/`, while in Huddle it uses `//`

But for the definition of the actual gov action it is defined as:
```cddl
gov_action =
  [ parameter_change_action
  // hard_fork_initiation_action
  // treasury_withdrawals_action
   ....
```
with usage of `//`, while for Huddle it is the opposite and `/` is used.

Moreover, to make things even more confusing `voting_procedure` uses a single `/` in both the original CDDL spec and in Huddle.


Could you please explain what is going on with those `/` vs `//` in Huddle and CDDL? My confusion probably come from my limited understanding of these operators in CDDL itself, but this inconsistency and lack of understanding is bothering me.

nc6 replied:
So, in CDDL, `/` is a choice between types (concretely, between `Type1` values, to make a `Type0`). `//` allows choice between groups. We can illustrate the difference:

```
foo = [ 0 / 1, uint // 2 /3, tstr ]
```
This construction would match either of the following:
```
[0, 3]
[2, "Hello World"]
```

 In other words, the `//` binds less strongly than `,` in CDDL.

In Haskell, of course, we cannot have syntax inside an array which binds stronger than the `,`. so we have to do things a little differently. The way this is handled at the moment is that `//` has special treatment for arrays/groups, where it will, instead of creating a type-level choice, merge the two arrays/groups/maps into a single one containing a group choice.

So, two questions remain: why is this operator `//` instead of `/` (whose semantics it more closely matches), and what is the `/` operator in Huddle?

The former is largely accidental: the group choice operator got expanded to other cases.

What about the latter? Well, that was mostly intended to handle the case of array entries. But looking at it again, I think it's a mistake. The case it was used for can be replaced by updating fixity declarations.

So I think I will update this with the following intention: all choices within huddle will be written with `/`, and I will incorporate most of this comment into the documentation to clarify it.

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,1008 @@
+          / update_drep_cert
+      ]
+
+stake_registration :: Named Group
```
Is there support in Huddle for comments?

It would be really useful to mark `stake_registration` and `stake_deregistration` as to  be deprecated in the future era.

nc6 replied:
Done

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,1008 @@
+      ]
+
+transaction_index :: Rule
+transaction_index = "transaction_index" =:= VUInt `sized` (2 :: Word64)
```
It would probably be useful to add `IsSize` instance for `Word` as well in Cuddle. But that is a minor suggestion for the future, rather that something we need to do right now
```suggestion
transaction_index = "transaction_index" =:= VUInt `sized` (2 :: Word)
```

nc6 replied:
https://github.com/input-output-hk/cuddle/issues/33

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,1008 @@
+      , opt $ idx 7 ==> nonempty_set plutus_v3_script
+      ]
+
+-- The real type of  plutus_v1_script, plutus_v2_script and plutus_v3_script is bytes.
+-- However, because we enforce uniqueness when many scripts are supplied,
+-- we need to hack around for tests in order to avoid generating duplicates,
+-- since the cddl tool we use for roundtrip testing doesn't generate distinct collections.
```
I think it is enough to have the same comment in Huddle.
```suggestion
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,1008 @@
+redeemers :: Rule
+redeemers =
+  "redeemers"
+    =:= sarr
```
Same here: adding a Huddle comment about deprecation would be very valuable

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,1008 @@
+script_n_of_k :: Named Group
+script_n_of_k =
+  "script_n_of_k"
+    =:~ grp [3, "n" ==> VUInt, a (arr [0 <+ a native_script])]
```
This is actually incorrect in the original spec. It is a signed 64bit `Int`. Could you please fix it here and in the spec. It's fine if you'd prefer not to change that in this PR then just feel free to convert this comment into a ticket and we'll take care of it later
```suggestion
    =:~ grp [3, "n" ==> int64, a (arr [0 <+ a native_script])]
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -0,0 +1,1008 @@
+pool_metadata_hash :: Rule
+pool_metadata_hash = "pool_metadata_hash" =:= hash32
+
+-- To compute a script hash, note that you must prepend
+-- a tag to the bytes of the script before hashing.
+-- The tag is determined by the language.
+-- The tags in the Conway era are:
+--   "\x00" for multisig scripts
+--   "\x01" for Plutus V1 scripts
+--   "\x02" for Plutus V2 scripts
+--   "\x03" for Plutus V3 scripts
```
Ditto
```suggestion
```

lehins reviewed (approved):
Few minor suggestions and question, but nothing critical.

Cuddle/Huddle is an amazing piece of work! Thank you for getting us to this point @nc6!
